plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs

```diff
@@ -0,0 +1,317 @@
+-- in the second.
+sizeSplit :: Int -> Int -> GenTm a -> (a -> GenTm b) -> GenTm (a, b)
+sizeSplit a b ga gb = do
+  n <- asks geSize
+  let na = (a * n) `div` (a + b)
+      nb = (b * n) `div` (a + b)
+  x <- withSize na ga
+  (,) x <$> withSize nb (gb x)
+
+-- * Names
+
+-- | Get all uniques we have generated and are used in the current context.
+getUniques :: GenTm (Set Unique)
+getUniques = do
+  GenEnv{geDatas = dts, geTypes = tys, geTerms = tms, geUnboundUsedTyNames = used} <- ask
+  return $ Set.mapMonotonic (_nameUnique . unTyName) (Map.keysSet dts <> Map.keysSet tys <> used) <>
+           Set.mapMonotonic _nameUnique (Map.keysSet tms) <>
+           Set.unions [ names d | d <- Map.elems dts ]
+  where
+    names (Datatype _ _ _ m cs) = Set.fromList $ _nameUnique m : [ _nameUnique c | VarDecl _ c _ <- cs ]
+
+{- Note [Warning about generating fresh names]: because `GenTm` is a *reader* monad
+   names are not immediately put into any state when generated. There is *no guarantee*
+   that in this situation:
+   ```
+   do nms <- genFreshNames ss
+      nms' <- genFreshNames ss
+   ```
+   the names in `nms` and `nms'` don't overlap.
+
+   Instead, what you are supposed to do is locally use the names in `nms` and `nms'` to
+   define generators that use them. This is done with functions like `bindTyName` and `bindTmName`:
+   ```
+   genLam ma mb = do
+      x <- genFreshName "x"
+      sizeSplit 1 7 (maybe (genType Star) return ma)
+                    --      v--- LOOK HERE!
+                    (\ a -> bindTmName x a . withNoEscape $ genTerm mb) $ \ a (b, body) ->
+                    --      ^--- LOOK HERE!
+                    TyFun () a b, LamAbs () x a body)
+   ```
+-}
+
+-- | Generate a fresh name. See Note [Warning about generating fresh names].
+genFreshName :: String -> GenTm Name
+genFreshName s = head <$> genFreshNames [s]
+
+-- | Generate one fresh name per string in the input list.
+-- names don't overlap. See Note [Warning about generating fresh names].
+genFreshNames :: [String] -> GenTm [Name]
+genFreshNames ss = do
+  used <- getUniques
+  let i = fromEnum $ Set.findMax $ Set.insert (Unique 0) used
+      js = [ j | j <- [1..i], not $ Unique j `Set.member` used ]
+      is = js ++ take (length ss + 10) [i+1..]
+  is' <- liftGen $ QC.shuffle is
+  return [Name (fromString $ s ++ show j) (toEnum j) | (s, j) <- zip ss is']
+
+-- | See `genFreshName`
+genFreshTyName :: String -> GenTm TyName
+genFreshTyName s = TyName <$> genFreshName s
+
+-- | See `genFreshNames`
+genFreshTyNames :: [String] -> GenTm [TyName]
+genFreshTyNames ss = map TyName <$> genFreshNames ss
+
+-- | Generate a name that likely overlaps with existing names on purpose. If there are no existing
+-- names, generate a fresh name. This function doesn't distinguish between the type- and term-level
+-- scopes, hence it may generate a 'Name' \"clashing\" with a previously generated 'TyName' and not
+-- clashing with any previously generated 'Name'.
```

Comment:
No, it's a good thing: we want to check that Plutus is able to handle such spurious name clashes. Generating weird stuff is useful for a testing machinery! We don't need any "definitely non-fresh name" anyway, since we get enough non-fresh names already and trying to add more logic to cover what is the already covered would only needlessly complicate things.

Reply:
Then maybe say that? As it is it sounds like this is a problem that you haven't fixed, whereas you're arguing that it's a feature! It also wasn't clear that you won't want "definitely non-fresh names", I guess you say that you don't really need them so it's okay, but if you're reliant on the ratio that comes out of this thing in order to make sure you're testing the non-fresh case that you sort-of ask for... seems simpler to just _ensure_ that it does the thing you wan :shrug:

Reply:
> Then maybe say that?

Done.

> but if you're reliant on the ratio

Who cares if it gets skewed slightly.

>  seems simpler to just ensure that it does the thing you wan shrug

Well, the thing is, it's not simpler with the current alignment. Otherwise sure, it'd be cleaner to have `genDefinitelyFreshName`, `genDefinitelyNonFreshName` and `genDefinitelyNonFreshInTheOtherScopeName`. Although it'd still be more logic and code. What we have right now is fine, we don't want to spend months here calculating the exact best possible ratio between fresh and non-fresh names to get a marginal coverage improvement when we can instead look into things on the critical path or indeed improve the generators machinery in places where it matters much more (such as faster fresh name generation, faster substitution, cleverer unification, statistical analysis of generated output etc).